xprocess.py

```diff
@@ -203,37 +203,60 @@ class ProcessStarter:
     before presuming the attached process dead.
     """
 
+    line_wait_timeout = 120
+    """The maximum time ProcessStarter.wait will hang waiting for a new
+    line when trying to match pattern before raising TimeoutError"""
+
     def __init__(self, control_dir, process):
         self.control_dir = control_dir
         self.process = process
 
     @abc.abstractproperty
     def args(self):
         "The args to start the process"
+        raise NotImplementedError
 
     @abc.abstractproperty
     def pattern(self):
         "The pattern to match when the process has started"
+        raise NotImplementedError
 
     def wait(self, log_file):
-        "Wait until the process is ready."
+        "Wait until the pattern is mached"
         lines = map(self.log_line, self.filter_lines(self.get_lines(log_file)))
         return any(std.re.search(self.pattern, line) for line in lines)
 
     def filter_lines(self, lines):
-        # only consider the first non-empty 50 lines
+        """fetch first <max_read_lines>, ignoring blank lines"""
         non_empty_lines = (x for x in lines if x.strip())
         return itertools.islice(non_empty_lines, self.max_read_lines)
 
     def log_line(self, line):
+        """Write line to process log file"""
         self.process.log.debug(line)
         return line
 
     def get_lines(self, log_file):
```

Comment:
This implementation looks good, but I would write it in terms of "elapsed" time rather than use a counter:

```python
    def get_lines(self, log_file):
        """Read and yield one line at a time from log_file. Will raise
        TimeoutError if a line read hangs for more than self.line_wait_timeout
        seconds"""
        waiting_line = None
        while True:
            print("** step counter: ", step_counter)
            line = log_file.readline()
            if not line:
                if waiting_line is None:
                    waiting_line = std.time.time()
                std.time.sleep(0.1)
                total_waiting = std.time.time() - waiting_line
                if total_waiting >= self.line_wait_timeout:
                    raise TimeoutError(
                        "The provided start pattern {} could not be matched \
                        within the specified time interval of {} seconds".format(
                            self.pattern, self.line_wait_timeout
                        )
                    )
            else:
                # reset counter
                waiting_line = None
            yield line
```

Totally up to you though, as both versions do the same. 👍

Reply:
Hm, yeah. I think I would also like your suggest approach more. I'll give this a second try...